### ISSUE TITLE
unstable backend: `best_block` or `all_blocks` subscription may be silently lose blocks when reconnecting

### ISSUE THREAD

**Issue by niklasad1** [paritytech/subxt#1568]
It's possible that blocks may be be silently missed if a reconnecting occurs because it's restarted by unstable backend.

For instance the following could happen:
1. A connection is established
2. Block A, Block B, and Block C are announced and propagated in the subscription
3. The connection is lost and during this phase Block D is announced
4. Then after a while the connection is re-established and the next Block E is announced.

Thus, Block D is missed and no way to now about it.

**Comment by jsdw** [comment]
Yeah, it's hard to know how much I care about missed new/best block notifications because they are ephemeral anyway; as long as all finalized blocks are notified then it's all good, and if you re-subscribe then you should be told about recent new blocks and the latest current best block again anyways :)

**Comment by lexnv** [comment]
Yep, something we can think about in the future.
One use-case may be users that want to query a pruned block, where block D was announced, the client disconnected, the chain pruned block D and the client reconnects.

**Comment by jsdw** [comment]
Right now we don't have a good solution for this. We use this RPC method (or the legacy equivalent) to resubscribe to blocks:

https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_v1_follow.html

This hands back a list of recently finalized blocks on `initialized`, and so we have a limited ability to pick up from where we left off w.r.t finalized blocks, but best blocks not so much; by the time you reconnect they may not be finalized, or may have been pruned, and we don't have any way to know (via the modern RPC, at least, which is what we are bound to).

> We've had to build our own replay mechanism for missed best blocks, so having it in subxt would simplify our code.

What approach did you take here?

**Comment by teor2345** [comment]
> > We've had to build our own replay mechanism for missed best blocks, so having it in subxt would simplify our code.
> 
> What approach did you take here?

Since we're building an alerts system, we are already tracking the previous best block hash, so we can compare that block with the next block. (This lets us check for blocks that take a long time, for example.)

If the next best block isn't a child of that block, we assume there's been a gap or reorg. Then we follow the parent block hashes back to the same hash/height as the previous block, storing those hashes. And we replay those blocks into the alerts system, from the start to the end of the gap.

This requires some additional RAM, but it's only 32 bytes per missed block.

**Comment by jsdw** [comment]
> Then we follow the parent block hashes back to the same hash/height as the previous block, storing those hashes. And we replay those blocks into the alerts system, from the start to the end of the gap.

For finalized blocks, we do something similar (but simpler) in that we just check whether the latest finalized block we knew about before disconnect is in the list of finalized blocks we are given on reconnect, and if so then we resume, and if not we return an error that we could not resume cleanly.

For best blocks, we apply the same logic and carry on from the last finalized block, but this ignored best blocks, since by the time we reconnect they may have been pruned or already finalized or whatever else. If pruned, we cannot access them and if finalized, we will see them in the stream anyway (along with the set of new blocks that exist on reconnect) :)
